### PR TITLE
Update systemboot for pxeboot and boot - developed on OCP Mono Lake

### PR DESCRIPTION
**Pull request by marcjones-syspro**
Issues and features found developing for OCP Mono Lake.

### CODE REVIEW

**Review by rminnich — APPROVED** [review]

**Comment by marcjones-syspro** [comment]
I'm not sure what to do about the codecov failures. If someone can direct me, I'd appreciate it. 

Also, this patch set needs comments from systemboot users. Do the pxeboot and boot booters meet the requirements to replace fbnetboot and localboot? What else is required before this lands?

**Comment by marcjones-syspro** [comment]
This is my first u-root patch set and I am not familiar with go or with github reviews and the u-root process. I could use some comments from @rminnich @hugelgupf @insomniacslk. Thanks!

**Comment by johnnylinwiwynn** [comment]
@rojer @tfg13 @jonzhang-fb

**Comment by marcjones-syspro** [comment]
How do I remove the waiting for author tag? How do I get reviews?

**Comment by johnnylinwiwynn** [comment]
Thanks for your work! Recently we plan to switch from 'cmds/boot/localboot' to 'cmds/boot/boot' since it's better maintained and used by the community, As described in https://github.com/opencomputeproject/OpenSystemFirmware/tree/master/Wiwynn/deltalake#potential-gaps-to-use-ocp-delta-lake-osf-in-production
The switch is pretty much done in your PR especially commit 'systemboot: Add pxeboot and boot booters'.
IPv4 and pxeboot probably are not used by Facebook, and I think we might ditch localboot. I cherry-pick your first 3 commits and it works. Let me see if we can help review your PR.

**Comment by insomniacslk** [comment]
I've completely missed this PR. @johnnylinwiwynn would that work for you if we review and merge this patch from @marcjones-syspro ?

**Comment by marcjones-syspro** [comment]
A review would be great! Any advice or other ideas about how I did this are welcome. Also any advice on tests coverage. Finally, I know that there are some conflicts since this branched 6 months ago, but getting a general idea on what other changes are needed would be helpful.

**Comment by johnnylinwiwynn** [comment]
> I've completely missed this PR. @johnnylinwiwynn would that work for you if we review and merge this patch from @marcjones-syspro ?

Yes please help to review. I will check the change in pkg/ipmi/ocp/bootorder.go which was added by me.
Thanks!

File: pkg/boot/systembooter/pxebooter.go
Comment (johnnylinwiwynn) [inline pkg/boot/systembooter/pxebooter.go:42]: From line 46 I think it should be pxeboot?

File: pkg/ipmi/ocp/bootorder.go
Comment (johnnylinwiwynn) [inline pkg/ipmi/ocp/bootorder.go:32]: For Facebook's use case I think it should be netboot (fbnetboot).
@insomniacslk is it okay to change from netboot to pxeboot for the default boot option? Although people can set their default booter configs using VPD at build time.

File: pkg/ipmi/ocp/bootorder.go
Comment (johnnylinwiwynn) [inline pkg/ipmi/ocp/bootorder.go:168]: This would replace netboot (fbnetboot) with pxeboot, I think we can keep netboot and add a new case for pxeboot.

File: pkg/ipmi/ocp/bootorder.go
Comment (johnnylinwiwynn) [inline pkg/ipmi/ocp/bootorder.go:103]: Rebasing to the current upstream this needs to be changed to
vpd.FlashromRWVpdSet(key, []byte(NETBOOTER_IPV6_CONFIG), false);

File: pkg/ipmi/ocp/bootorder.go
Comment (johnnylinwiwynn) [inline pkg/ipmi/ocp/bootorder.go:127]: I think BmcUpdatedBootorder should be renamed to VpdUpdatedBootorder since it could be updated both by BMC and BIOS. On line 39 can be changed to:
// VpdUpdatedBootorder is true when VPD boot order has been updated
VpdUpdatedBootorder bool = false

File: pkg/boot/systembooter/bootbooter.go
Comment (johnnylinwiwynn) [inline pkg/boot/systembooter/bootbooter.go:59]: Should be 'boot' command

File: cmds/boot/systemboot/main.go
Comment (johnnylinwiwynn) [inline cmds/boot/systemboot/main.go:51]: @insomniacslk should we remove localboot from the hard-coded defaultBootsequence or keep it around? And do you have any preference about the boot order sequence here for Facebook's use case?

File: pkg/ipmi/ocp/bootorder.go
Comment (marcjones-syspro) [reply pkg/ipmi/ocp/bootorder.go:32]: I considered other ways to do this, but it seemed that the most generic way was best. The problem is VPD vs IPMI.

File: pkg/ipmi/ocp/bootorder.go
Comment (marcjones-syspro) [reply pkg/ipmi/ocp/bootorder.go:103]: ack

File: pkg/ipmi/ocp/bootorder.go
Comment (marcjones-syspro) [reply pkg/ipmi/ocp/bootorder.go:168]: The problem is that there is only one type of network boot in IPMI boot message.

File: pkg/ipmi/ocp/bootorder.go
Comment (marcjones-syspro) [reply pkg/ipmi/ocp/bootorder.go:127]: I agree. It has been 6 months, but I found a hole in this logic when either the BIOS or IPMI updated the boot option and the VPD needed to be updated. As you see here the VPD is always reordered when entering updateVPDBootOrder, so the flag moved here.

File: cmds/boot/systemboot/main.go
Comment (marcjones-syspro) [inline cmds/boot/systemboot/main.go:345]: I'm not sure what I mean here. This case probably needs testing.

File: pkg/ipmi/ocp/bootorder.go
Comment (johnnylinwiwynn) [reply pkg/ipmi/ocp/bootorder.go:168]: I think we can add one VPD variable 'pxeboot' for selecting pxeboot or fbnetboot. I made a commit for your reference
https://github.com/johnnylinwiwynn/u-root/commit/12555d5192ff0170cac7df518aebeefca7ef629e
@insomniacslk

File: pkg/ipmi/ocp/bootorder.go
Comment (johnnylinwiwynn) [reply pkg/ipmi/ocp/bootorder.go:32]: Same as my comment in https://github.com/u-root/u-root/pull/2045#discussion_r716434905

Comment (marcjones-syspro) [reply pkg/ipmi/ocp/bootorder.go:32]: > Same as my comment in [#2045 (comment)](https://github.com/u-root/u-root/pull/2045#discussion_r716434905)

Seems reasonable if fbnetboot is needed. Can pxeboot in ipv6 mode be a full replacement?

Comment (johnnylinwiwynn) [reply pkg/ipmi/ocp/bootorder.go:32]: > > Same as my comment in [#2045 (comment)](https://github.com/u-root/u-root/pull/2045#discussion_r716434905)
> 
> Seems reasonable if fbnetboot is needed. Can pxeboot in ipv6 mode be a full replacement?

@insomniacslk what do you think?

**Comment by rminnich** [comment]
marc, let's revive this.

**Comment by rminnich** [comment]
Please note that I closed #2151, take a look and you can see why. 
https is supported, but you need an /etc/ssl/certs directory, e.g. add this to your u-root command
```
-files /etc/ssl/certs
```

File: pkg/ipmi/ocp/bootorder.go
Comment (rminnich) [reply pkg/ipmi/ocp/bootorder.go:33]: we need one, but I've never figured out how it should look. possibly the osf builder is the toold.

**Comment by rminnich** [comment]
I am ready to approve this if I can figure out if all the people commenting on it are ok with it?

**Comment by rminnich** [comment]
for the record, this is a gigantic patch and it would be best in future if they could be smaller. I got totally lost trying to review it.

**Comment by johnnylinwiwynn** [comment]
> 

Please hold on merging this, @jonzhang-fb, @marcjones-syspro we will have the chance to discuss this in the near future.
Thanks.

**Comment by marcjones-syspro** [comment]
> for the record, this is a gigantic patch and it would be best in future if they could be smaller. I got totally lost trying to review it.

Ron, Thanks for the review. I agree it was a large change, but I didn't see a good place to break it. I had hoped for some comments while it was being developed and I could have made adjustments that would make it easier to review. 

 I guess we are waiting to review this with those that need fbnetboot.

**Comment by rjoleary** [comment]
A big piece of this PR is adding the menuName. Can you break it into a separate PR? Since the menu name is mainly cosmetic, we can get it reviewed faster.

**Comment by rminnich** [comment]
this commit is almost two years old, we need to figure out what to do with it.

**Comment by rminnich** [comment]
oh woops only one year old but still

**Comment by marcjones-syspro** [comment]
I think that the systemboot stakeholders need to decide to move from fbnetboot to pxeboot. 

As for this being a big change, I don't know the correct way to handle dependencies, as we do on coreboot with gerrit.

**Comment by rminnich** [comment]
This should be two PRs, as Ryan indicated; it has been hanging for a very long time; and we have no idea if anyone even needs it any more. It can be revived, as several PRs, but not in this form.

**Pull request closed without merging**
